src/hades/leases/server.py

```diff
@@ -0,0 +1,515 @@
+                data.seek(end + 1, os.SEEK_SET)
+                return data, size, arg
+            else:
+                need = size - start + 1
+
+    @classmethod
+    def parse_request(
+            cls,
+            data: mmap.mmap,
+            size: int,
+    ) -> Parser[Tuple[List[bytes], Dict[bytes, bytes]]]:
+        # Parse number of arguments
+        element = "argc"
+        data, size, argc = yield from cls.parse_int(data, size, element)
+        if argc < 0:
+            raise ParseError("Negative argc: " + str(argc), element=element)
+
+        # Parse arguments
+        argv = []
+        for i in range(argc):
+            element = "argv[{:d}]".format(i)
+            data, size, arg = yield from cls.parse_string(data, size, element)
+            argv.append(arg)
+
+        # Parse number of environment variables
+        element = "envc"
+        data, size, envc = yield from cls.parse_int(data, size, element)
+        if envc < 0:
+            raise ParseError("Negative envc: " + str(envc), element=element)
+
+        # Parse environment variables
+        environ = {}
+        for i in range(envc):
+            element = "environ[{}]".format(i)
+            data, size, env = yield from cls.parse_string(data, size, element)
+            name, sep, value = env.partition(b'=')
+            if not sep:
+                raise ParseError(
+                    "No equal sign in environment variable: " + repr(name),
+                    element=element,
+                )
+            environ[name] = value
+        return data, size, (argv, environ)
+
+    def _handle_shutdown_signal(self, signo, _frame):
+        logger.error("Received signal %d. Shutting down.", signo)
+        # shutdown blocks until the server is stopped, therefore we must use a
+        # separate thread, otherwise there will be deadlock
+        threading.Thread(name='shutdown', target=self.shutdown).start()
+
+    def serve_forever(self, poll_interval=0.5):
+        logger.info("Starting server loop")
+        with install_handler(
+            (signal.SIGHUP, signal.SIGINT, signal.SIGTERM),
+            self._handle_shutdown_signal
+        ):
+            super().serve_forever(poll_interval)
+
+    def _process(
+            self,
+            stdin: TextIO, stdout: TextIO, stderr: TextIO,
+            args: Sequence[bytes], env: Dict[bytes, bytes]
+    ) -> int:
+        decoded_args = [decode(a) for a in args]
+        parsed_args = self.parser.parse_args(decoded_args[1:])
+        return dispatch_commands(
+            args=parsed_args,
+            context=Context(
+                stdin=stdin, stdout=stdout, stderr=stderr,
+                environ={decode(k): decode(v) for k, v in env.items()},
+                environb=env,
+            ),
+            engine=self.engine,
+        )
+
+
+def decode(x: bytes) -> str:
+    """Decode a string like done in `os._createenviron` (hard-coding utf-8)"""
+    return x.decode("utf-8", errors="surrogateescape")
```

Comment:
Why not `sys.getfilesystemencoding()`? That's also what `os._createenviron` does.
```suggestion
    return x.decode(sys.getfilesystemencoding(), errors="surrogateescape")
```